Add Blast Block Modification Event

AFlyingCar opened the pull request:
Implemented a new `BlastBlockModifyEvent` and `BlastBlockModifyHandler` for moving block destruction code out of the various Blast implementations. This allows block destruction to be interrupted by other mods, allowing block/chunk protection to be implemented externally through the use of mods such as Clans, addressing many of the points in issue #201 .

AFlyingCar committed:
Added an event and handler for blasts breaking blocks. Modified every blast to fire this event rather than setting block states directly.

AFlyingCar committed:
Fixed posted event for antimatter to not post with a lambda, but to post a blockstate  instead.

AFlyingCar committed:
Renamed BlastBreakEvent and BlastBreakHandler to BlastBlockModifyEvent and BlastBlockModifyHandler to more accurately represent what they are meant for.

DarkGuardsman commented:
I'll review this when I have time later this week. For the moment do not merge it @GHXX

DarkGuardsman commented on src/main/java/icbm/classic/api/events/BlastBlockModifyEvent.java:
```diff
@@ -0,0 +1,93 @@
+import net.minecraft.world.World;
+import net.minecraftforge.fml.common.eventhandler.Event;
+
+public class BlastBlockModifyEvent extends Event
```
Missing documentation

DarkGuardsman commented on src/main/java/icbm/classic/api/events/BlastBlockModifyEvent.java:
```diff
@@ -0,0 +1,93 @@
+    private Runnable callback = null;
+    private boolean placedBackDown = false;
+
+    public enum BlockBreakType
```
I rather not have nested classes in my APIs

AFlyingCar replied:
I can go ahead and pull `BlockBreakType` out of `BlastBlockModifyEvent` (as well as renaming it so that it's still obvious what it is meant to be used with), however is there a specific place in `api` that I should place the new file? Would you prefer for me to place it in `events` with `BlastBlockModifyEvent`, or should I create a new package for it?

DarkGuardsman commented on src/main/java/icbm/classic/api/events/BlastBlockModifyEvent.java:
```diff
@@ -0,0 +1,93 @@
+
+    private BlockBreakType modificationType;
+
+    public BlastBlockModifyEvent(World _world, BlockPos _position)
```
Change _argName to argName. Then do `this.world = world`

DarkGuardsman commented on src/main/java/icbm/classic/content/blast/BlastAntiGravitational.java:
```diff
@@ -88,26 +90,30 @@ public boolean doExplode(int callCount) //TODO rewrite entire method
                                     {
-                                        //Remove block
-                                        world.setBlockToAir(targetPosition);
+                                        MinecraftForge.EVENT_BUS.post(new BlastBlockModifyEvent(world, targetPosition, () -> {
```
I rather not have large lambda statements. Convert to method or function.

DarkGuardsman commented on src/main/java/icbm/classic/content/blast/BlastBlockModifyHandler.java:
```diff
@@ -0,0 +1,45 @@
+
+        switch(event.getModificationType()) {
+            case SET_TO_AIR:
+                // System.out.println("Set To Air");
```
Don't leave commented out code in a PR

DarkGuardsman commented on src/main/java/icbm/classic/content/blast/BlastBlockModifyHandler.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.function.Predicate;
+
+@Mod.EventBusSubscriber
+public class BlastBlockModifyHandler {
```
This whole class is the wrong way to do events in forge.

AFlyingCar replied:
Out of curiosity, are you referring to the use of the `Predicate` callback? Or the missing `modid = ICBMConstants.DOMAIN`? I have fixed the missing `modid = ICBMConstants.DOMAIN` and can remove the callback part fairly easily.

DarkGuardsman replied:
callback and the use of an event to capture event calls. Rather than letting the base code handle a lack of event capture. Event system mostly is built to work in a way that you fire the event, some mod changes the data in the event, and then the base code gets the changes back. Other option is the event is fired, a mod cancels the event, and then the mod does something different.

DarkGuardsman reviewed (changes requested):
Did a quick review, will do a more detailed review after changes are made or I get time to pull the code locally. I should be able to find some time as my job just went into e2e testing. So I will not be working 10 hour days this week.

AFlyingCar committed:
Addressed requested changes.

Added documentation to BlastBlockModifyEvent. Broke large lambdas up into smaller methods. Changed _argName parameters to argName. Removed commented out code. Broke BlastBlockModifyEventType into a seperate file. Removed predicate callback from BlastBlockModifyHandler. Added Objects.requreNonNull calls.

AFlyingCar committed:
Changed flying block placement to also fire an event.
